models/lenet/lenet_impl.hpp

```diff
@@ -0,0 +1,105 @@
+/**
+ * @file lenet_impl.hpp
+ * @author Eugene Freyman
+ * @author Daivik Nema
+ * @author Kartik Dutt
+ * 
+ * Implementation of LeNet using mlpack.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_LENET_IMPL_HPP
+#define MODELS_LENET_IMPL_HPP
+
+#include "lenet.hpp"
+namespace mlpack {
+namespace ann {
+
+template<size_t leNetVer>
+LeNet<leNetVer>::LeNet(const size_t inputChannel,
+                       const size_t inputWidth,
+                       const size_t inputHeight,
+                       const size_t numClasses,
+                       const std::string& weights) :
+                       LeNet<leNetVer>(
+                          std::tuple<size_t, size_t, size_t>(
+                              inputChannel,
+                              inputWidth,
+                              inputHeight),
+                          numClasses,
+                          weights)
+{
+  // Nothing to do here.
+}
+
+template<size_t leNetVer>
+LeNet<leNetVer>::LeNet(const std::tuple<size_t, size_t, size_t> inputShape,
+                       const size_t numClasses,
+                       const std::string &weights) :
+                       inputChannel(std::get<0>(inputShape)),
+                       inputWidth(std::get<1>(inputShape)),
+                       inputHeight(std::get<2>(inputShape)),
+                       numClasses(numClasses),
+                       weights(weights)
+{
+  leNet = new Sequential<>();
+  ConvolutionBlock(inputChannel, 6, 5, 5, 1, 1, 2, 2);
+  PoolingBlock(2, 2, 2, 2);
+  ConvolutionBlock(6, 16, 5, 5, 1, 1, 2, 2);
+  PoolingBlock(2, 2, 2, 2);
+  // Add linear layer for LeNet.
+  if (leNetVer == 1)
+  {
+    leNet->Add<Linear<>>(16 * inputWidth * inputHeight, numClasses);
+  }
+  else if (leNetVer == 4)
+  {
+    leNet->Add<Linear<>>(16 * inputWidth * inputHeight, 120);
+    leNet->Add<LeakyReLU<>>();
+    leNet->Add<Linear<>>(120, numClasses);
+  }
+  else if (leNetVer == 5)
+  {
+    leNet->Add<Linear<>>(16 * inputWidth * inputHeight, 120);
+    leNet->Add<LeakyReLU<>>();
+    leNet->Add<Linear<>>(120, 84);
+    leNet->Add<LeakyReLU<>>();
+    leNet->Add<Linear<>>(84, 10);
+  }
```

Comment:
If `leNetVer` is a template parameter, it might make the most sense to simply split these out into separate functions, like:

```
template<>
LeNet<1>::BuildNetwork()
{
  leNet->Add<Linear<>>(16 * inputWidth * inputHeight, numClasses);
}

template<>
LeNet<4>::BuildNetwork()
{
  leNet->Add<Linear<>>(16 * inputWidth * inputHeight, 120);
  leNet->Add<LeakyReLU<>>();
  leNet->Add<Linear<>>(120, numClasses);
}

...
```

Then you can just call `BuildNetwork()` from this constructor and it will select the right specialization. :)